### PR TITLE
Reporter receive notification when someone comments on their issue

### PR DESCRIPTION
**Pull request by wijayaerick**
#### Summary

This PR modifies `appendCommentNotifications` so both Assignee and Reporter will get a notification when someone comments on their issue (previously, only Assignee gets the notification)

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-jira/issues/699

### CODE REVIEW

File: server/webhook_parser.go
Comment (levb) [inline server/webhook_parser.go:324]: nit: 0/5 `s/can/should/`?

**Review by levb — APPROVED** [review]
One small nit/style suggestion, and one wish... can you add a unit test that covers this logic? We've had snafus in the past with who gets notified.

**Comment by wijayaerick** [comment]
Thank you for the review @levb. For the unit test, should I write a new test function `TestXX`, or perhaps there're already existing test tables that I can extend?

File: server/webhook_parser.go
Comment (jfrerich) [inline server/webhook_parser.go:301]:
```suggestion
	// Don't send a notification to the assignee or reporter if they don't exist, or if they are also the author.
```

**Review by jfrerich — APPROVED** [review]
Thanks for the PR, @wijayaerick!  Your code looks good to me, but would you please fix the grammar suggestion I found.  This wasn't related to anything you did, but would be great if you could fix it for us in this PR. :)

**Comment by levb** [comment]
@wijayaerick Thank you!
@aaronrothschild Can we please get PM approval, this is a very useful change.

File: server/webhook_parser_misc_test.go
Comment (levb) [inline server/webhook_parser_misc_test.go:218]: 👍 thank for adding these!

**Review by aaronrothschild — APPROVED** [review]
LGTM, this is great!

**Comment by aaronrothschild** [comment]
@justinegeffen Could you add this to the documentation about the notifications sent to users by the plugin?

**Comment by DHaussermann** [comment]
@aaronrothschild  I'll be testing this shortly. Just something to point out about this feature, for some users who report many issues (which can stay active a long time) this could make the Jira DM bot very noisy. 

We should consider some type of configuration to accompany this feature (possibly out of scope for this PR). I would be ideal if the user had the power to disable some types of notifications.

**Comment by DHaussermann** [comment]
@wijayaerick Thanks for this PR. I have done a round of testing and this seems to be working on Jira Server. Comments are delivered when the connected user is the reporter and only when comments are posted by other users 🎉 

I notice on Jira Cloud that I don't see these DMs being delivered. I'm not sure if there is a change needed to your PR or the problem is elsewhere. Note that there has been an issue on Cloud https://github.com/mattermost/mattermost-plugin-jira/issues/479 with closely related functionality where even the assignee is not getting DMs.

**Comment by aaronrothschild** [comment]
> @aaronrothschild I'll be testing this shortly. Just something to point out about this feature, for some users who report many issues (which can stay active a long time) this could make the Jira DM bot very noisy.
> 
> We should consider some type of configuration to accompany this feature (possibly out of scope for this PR). I would be ideal if the user had the power to disable some types of notifications.

@wijayaerick - @DHaussermann brings up a good point here that I'm sorry I didn't think of earlier.  There are likely some users who are assigned tickets by default as part of a process, etc. that could overwhelm them with DM notifications and reduce the utility of the plugin for them.  We can open another ticket or we can address it within this PR, but I don't think we can release this functionality without a notification settings flag such as "send_notifications_for_assigned" and "send_notifications_for_reporter".   which would control if the Jira user wants to get these new notifications - or If they should be disabled. Adding those options to `/jira settings` as options would work. Thoughts?

**Comment by jasonblais** [comment]
@DHaussermann friendly reminder to help with review :)

**Comment by DHaussermann** [comment]
Thanks for pinging this back onto the radar @jasonblais I did test this PR and to clarify my understanding of where this is at:

1. There is an issue where this new feature is working on Jira Cloud but not Jira server. (may be related to an existing bug https://github.com/mattermost/mattermost-plugin-jira/issues/479)

2. Do we need a mechanism for users to toggle the new notifications on and off before we merge this. 

cc @wijayaerick and @aaronrothschild

**Comment by wijayaerick** [comment]
I think we'll need to address the toggle notifications settings in this PR, knowing that we cannot merge this PR without it.

**Comment by jfrerich** [comment]
@wijayaerick, would you be willing to add the setting into this PR?  If you don't have time, we can open an issue for that request and reference it in this PR but it would definitely add time to getting your PR merged.

**Comment by wijayaerick** [comment]
Hi @jfrerich sorry for the inactivity
I'm planning to add the setting into this PR, hopefully by the end of this week

File: server/settings.go
Comment (levb) [inline server/settings.go:13]: nit: 0/5 an (inline-able) function with a switch statement is slightly preferred. Better performance (not that it matters here) and a tad easier on the eye?

**Review by levb — APPROVED** [review]
Mostly good, need to resolve the conflicts.

**Review by jfrerich — APPROVED** [review]
LGTM

**Comment by mickmister** [comment]
@wijayaerick Thanks for your contribution! This really is an awesome feature :+1:

Sorry for having this PR open for so long. As it stands, it still needs to be QA reviewed.

I have one request going forward with the PR. Please avoid force-pushing, as it re-pushes all commits, so we have to review the whole PR again instead of where we left off with our previous review. If there are ever merge conflicts to resolve, it's recommended to merge master into your PR's branch, rather than rebasing and force pushing.

Let me know if you have any questions about this. Mattermost developer Jesse Hallam made a blog article that mentions this https://mattermost.com/blog/submitting-great-prs

File: server/webhook_parser.go
Comment (mickmister) [inline server/webhook_parser.go:320]: It looks like if the reporter is the assignee, they will receive two notifications. Can we avoid this somehow?

**Review by mickmister — CHANGES_REQUESTED** [review]
I see some breaking UX changes in this PR. Thee previous logic was:

if `notifications` setting is true, send mentions and assignee notifications

Now since we've introduced two additional notification settings, `assignee` and `reporter` (both off my default), the logic is a bit different:

if `notifications` setting is true:
    - send mention notifications
    - if `assignee_notifications` is true (default is false), send notification if user is the assignee
    - if `reporter_notifications` is true (default is false), send notification if user is the reporter

@DHaussermann This is a break in UX, as if I were subscribing to notifications, I need to manually update my settings in order to continue receiving notifications for as an assignee. This will not be obvious to users.

To circumvent this, I suggest we store the settings as strings rather than booleans, and if the setting is blank, we assume that it is set to `on`. Another solution would be to store an inverted setting like `AvoidAssigneeNotifications`.

Another thing I noticed: In different areas of the code where we are initializing a user's struct when they connect, we usually set defaults for the notification settings. But this PR does not add defaults for these new settings. @wijayaerick Could you please add the default values in places like this:

https://github.com/mattermost/mattermost-plugin-jira/blob/c8ab47b5d44ded2d9d4ab29876dc8b8f1a263102/server/user_cloud.go#L103-L107

File: server/webhook_parser.go
Comment (mickmister) [inline server/webhook_parser.go:325]: To fix the issue with Jira cloud, we need to add an extra check. I suggest making these a little more verbose since we made it its own function:

```suggestion
    if to == nil {
        return false
    }
    
    // check for Jira Server user
    if to.Name != "" && to.Name == fromName {
        return false
    }    
    
    // check for Jira Cloud user
    if to.AccountID != "" && to.AccountID == fromAccountID {
        return false
    }
    
    return true
}
```

The check for `to.Name != ""` is the actual fix here. With this fix, assignees and reporters should get DM notifications on Jira Cloud. cc @DHaussermann

File: server/user.go
Comment (mickmister) [inline server/user.go:50]: To make things backwards compatible, we can introduce `*bool` fields. This allows us to know if the value should be in its "default" state.

```suggestion
    Notifications bool `json:"notifications"`
	SendNotificationsForMention *bool `json:"send_notifications_for_mention"`
	SendNotificationsForAssignee *bool `json:"send_notifications_for_assignee"`
    SendNotificationsForReporter *bool `json:"send_notifications_for_reporter"`
```

Then we can use this logic to determine the value:

```go
sendForAssigned := false
if settings.SendNotificationsForAssigned != nil {
    sendForAssigned = *settings.SendNotificationsForAssigned
} else {
    sendForAssigned = settings.Notifications
}
```

The user then sees three notification options available when changing their settings. When a user connects, they should have the default settings of all notification values as true.

**Comment by mickmister** [comment]
@aaronrothschild Currently there is only one notification setting the user can control. The one setting controls both mentions and assignee notifications. I'm thinking we should assume that if the user currently has notifications on, that they should receive reporter notifications once they have the feature. They will have the option of toggling mention, assignee, and reporter separately in this version.

**Comment by mickmister** [comment]
Hi @wijayaerick! How is everything going?

**Comment by maisnamrajusingh** [comment]
@mickmister would you like us to take over and make the remaining changes ?

**Comment by mickmister** [comment]
Hi @wijayaerick, if you'd like we can pick up the PR as it currently stands. Please let me know your thoughts within the next few days. Thanks for your contribution @wijayaerick!

cc @maisnamrajusingh

**Comment by wijayaerick** [comment]
Hi @mickmister @maisnamrajusingh, please take over the PR as it currently stands. I'm sorry I didn't make the remaining changes.

**Comment by mickmister** [comment]
No worries @wijayaerick! Thanks again for this great contribution!!

**Comment by DHaussermann** [comment]
There is not enough time to finish dev and testing on this for the `v3.1.0` comming up Soon. I have removed the milestone. 
cc @aaronrothschild

**Comment by maisnamraju** [comment]
assigning this to @sibasankar-demansol

**Comment by sibasankarnayak** [comment]
> assigning this to @sibasankar-demansol

have took the base code from this PR and raised the PR https://github.com/mattermost/mattermost-plugin-jira/pull/805 

CC @hanzei  @mickmister

**Comment by mickmister** [comment]
Closing in favor of https://github.com/mattermost/mattermost-plugin-jira/pull/805

Thanks again for your great work on this @wijayaerick!! This is feature is going to be awesome when it's released!

**Pull request closed without merging**
